src/file_scanner/arrow_multi_file_info.cpp

```diff
@@ -0,0 +1,244 @@
+
+//! The Arrow Local File State, basically refers to the Scan of one Arrow File
+//! This is done by calling the Arrow Scan directly on one file.
+struct ArrowFileLocalState : public LocalTableFunctionState {
+ public:
+  ArrowFileLocalState(ExecutionContext& execution_context)
+      : execution_context(execution_context) {};
+  //! Factory Pointer
+  shared_ptr<ArrowFileScan> file_scan;
+
+  ExecutionContext& execution_context;
+
+  //! Each local state refers to an Arrow Scan on a local file
+  unique_ptr<ArrowScanFunctionData> local_arrow_function_data;
+  unique_ptr<TableFunctionInitInput> init_input;
+  unique_ptr<GlobalTableFunctionState> local_arrow_global_state;
+  unique_ptr<LocalTableFunctionState> local_arrow_local_state;
+  unique_ptr<TableFunctionInput> table_function_input;
+};
+
+unique_ptr<LocalTableFunctionState> ArrowMultiFileInfo::InitializeLocalState(
+    ExecutionContext& context, GlobalTableFunctionState& function_state) {
+  return make_uniq<ArrowFileLocalState>(context);
+}
+
+shared_ptr<BaseFileReader> ArrowMultiFileInfo::CreateReader(
+    ClientContext& context, GlobalTableFunctionState& gstate_p, BaseUnionData& union_data,
+    const MultiFileBindData& bind_data) {
+  return make_shared_ptr<ArrowFileScan>(context, union_data.GetFileName());
+}
+
+shared_ptr<BaseFileReader> ArrowMultiFileInfo::CreateReader(
+    ClientContext& context, GlobalTableFunctionState& gstate_p, const string& filename,
+    idx_t file_idx, const MultiFileBindData& bind_data) {
+  return make_shared_ptr<ArrowFileScan>(context, filename);
+}
+
+shared_ptr<BaseFileReader> ArrowMultiFileInfo::CreateReader(
+    ClientContext& context, const string& filename, ArrowFileReaderOptions& options,
+    const MultiFileOptions& file_options) {
+  return make_shared_ptr<ArrowFileScan>(context, filename);
+}
+
+shared_ptr<BaseUnionData> ArrowMultiFileInfo::GetUnionData(
+    shared_ptr<BaseFileReader> scan_p, idx_t file_idx) {
+  auto& scan = scan_p->Cast<ArrowFileScan>();
+  auto data = make_shared_ptr<BaseUnionData>(scan_p->GetFileName());
+  if (file_idx == 0) {
+    data->names = scan.GetNames();
+    data->types = scan.GetTypes();
+    data->reader = std::move(scan_p);
+  } else {
+    data->names = scan.GetNames();
+    data->types = scan.GetTypes();
+  }
+  return data;
+}
+
+void ArrowMultiFileInfo::FinalizeReader(ClientContext& context, BaseFileReader& reader,
+                                        GlobalTableFunctionState&) {}
+
+bool ArrowMultiFileInfo::TryInitializeScan(ClientContext& context,
+                                           shared_ptr<BaseFileReader>& reader,
+                                           GlobalTableFunctionState& gstate_p,
+                                           LocalTableFunctionState& lstate_p) {
+  auto& gstate = gstate_p.Cast<ArrowFileGlobalState>();
+  auto& lstate = lstate_p.Cast<ArrowFileLocalState>();
+  if (gstate.files.find(reader->file_list_idx.GetIndex()) != gstate.files.end()) {
+    return false;
```

Comment:
With this you're essentially saying to the MultiFileReader: "only one thread is allowed to scan this file"
This is the place where you can handle parallelization over a single file, for reference this is what parquet does:
```c++
	if (gstate.row_group_index >= reader.NumRowGroups()) {
		// scanned all row groups in this file
		return false;
	}
```